src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -5262,97 +5027,121 @@ static bool INumberBase<BigInteger>.TryConvertToTruncating<TOther>(BigInteger va
                 // After a 32 bit right shift, it becomes [0x00, 0x00] which is [0x00, 0x00] when converted back.
                 // The expected result is [0x00, 0x00, 0xFFFFFFFF] (2's complement) or [0x00, 0x00, 0x01] when converted back
                 // If the 2's component's last element is a 0, we will track the sign externally
-                ++xl;
-            }
-
-            uint[]? xdFromPool = null;
-            Span<uint> xd = (xl <= BigIntegerCalculator.StackAllocThreshold
-                          ? stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
-                          : xdFromPool = ArrayPool<uint>.Shared.Rent(xl)).Slice(0, xl);
-            xd[^1] = 0;
-            bits.CopyTo(xd);
+                ++zLength;
 
-            if (negx)
-            {
-                if (shiftAmount >= ((long)kcbitUint * xd.Length))
+                if (bits[^1] == kuMaskHighBit && negLeadingZeroCount == bits.Length - 1)
                 {
-                    result = MinusOne;
-                    goto exit;
+                    // When bits are [0, ..., 0, 0x80000000], special handling is required.
+                    // Since the bit length remains unchanged in two�fs complement, the result must be computed directly.
+                    --zLength;
+                    if (zLength <= 0)
+                        return s_bnMinusOneInt;
+
+                    if (zLength == 1)
+                        return new BigInteger(int.MinValue >>> smallShift);
+
+                    uint[] rgu = new uint[zLength];
+                    rgu[^1] = kuMaskHighBit >>> smallShift;
+                    return new BigInteger(smallShift == 0 ? -1 : +1, rgu);
                 }
-
-                NumericsHelpers.DangerousMakeTwosComplement(xd); // Mutates xd
             }
+            else if (zLength <= 0)
+            {
+                goto Excess;
+            }
+
+            uint[]? zFromPool = null;
+            Span<uint> zd = ((uint)zLength <= BigIntegerCalculator.StackAllocThreshold
+                            ? stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                            : zFromPool = ArrayPool<uint>.Shared.Rent(zLength)).Slice(0, zLength);
 
-            uint[]? zdFromPool = null;
-            int zl = Math.Max(xl - digitShift, 0);
-            Span<uint> zd = ((uint)zl <= BigIntegerCalculator.StackAllocThreshold
-                          ? stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
-                          : zdFromPool = ArrayPool<uint>.Shared.Rent(zl)).Slice(0, zl);
-            zd.Clear();
+            zd[^1] = 0;
+            bits.Slice(digitShift).CopyTo(zd);
 
-            if (smallShift == 0)
+            if (neg)
             {
-                for (int i = xd.Length - 1; i >= digitShift; i--)
+                // Calculate the two�fs complement. The least significant nonzero bit has already been computed.
```

Comment:
non-ASCII character

Reply:
I had used U+2019 `’`, so I replaced it with U+0027 `'`.